### PR TITLE
Adding Typescript implementation of the tiktoken algorithm.

### PR DESCRIPTION
**Pull request by shengyfu**
1. Adding Typescript implementation of the tiktoken algorithm.
2. Minor clean up on the C# version.

### CODE REVIEW

File: tokenizer_ts/test/tikTokenizer_p50k_base.test.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    [IM_END, 100265],
+]);
+
+suite('TikTokenizer p50k_base Test Suite', function () {
+    let tokenizer_p50k_base: TikTokenizer;
+    beforeEach(async () => {
+        tokenizer_p50k_base = await createByEncoderName("p50k_base", specialTokens);
+    });
+
+    test('tokenize source code - p50k_base', () => {
```
Comment (jinu-jang) [inline tokenizer_ts/test/tikTokenizer_p50k_base.test.ts:24]: I'm surprised this works. However we can be more explicit in our wanted behavior.

1. Pass in done() to let Mocha know exactly when the callbacks are done:
```suggestion
    test('tokenize source code - p50k_base', (done) => {
```

```typescript
fs.readfile(filePath, 'utf8', (err, data) => {
    if (err) {
        done(err);
    }
    ...
    done();
```

2. Just use readFileSync
```typescript
const fileContent = fs.readFileSync(filePath, 'utf8')
```

File: tokenizer_ts/test/tikTokenizer_p50k_base.test.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        let source = fs.readFileSync('test/testdata/lib.rs.txt', 'utf8');
```
Comment (jinu-jang) [inline tokenizer_ts/test/tikTokenizer_p50k_base.test.ts:25]:
```suggestion
        const source = fs.readFileSync('test/testdata/lib.rs.txt', 'utf8');
```

File: tokenizer_ts/test/tikTokenizer.test.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  });
+
+  test('special tokens - 2', () => {
+    const str = "<|im_start|>Hello<|im_end|> World";
+    let encoded = tokenizer.encode(str, Array.from(specialTokens.keys()));
+    assert.deepStrictEqual(encoded, [100264, 9906, 100265, 4435]);
+    assert.strictEqual(tokenizer.decode(encoded), str)
+  });
+
+  test('special tokens - 3', () => {
```
Comment (jinu-jang) [inline tokenizer_ts/test/tikTokenizer.test.ts:59]:
```suggestion
  test('special tokens with unicode', () => {
```

File: tokenizer_ts/test/testdata/tokens.json [hunk 4]
```diff
@@ -0,0 +1 @@
...9, 262, 557, 262, 443, 25855, 20168, 262, 443, 79377, 198, 262, 443, 25855, 8880, 271, 262, 5279, 4037, 20317, 9324, 2146, 726, 11, 4611, 25, 13325, 8, 1492, 11564, 27, 14149, 27, 14149, 7223, 2511, 341, 286, 659, 70197, 6594, 12807, 198, 310, 662, 2058, 746, 310, 662, 2235, 23236, 87, 91, 5468, 7223, 487, 943, 47927, 11, 865, 570, 18614, 2455, 310, 662, 17840, 746, 262, 457, 633, 13657, 79, 1631, 1793, 933, 8998, 721, 83, 1609, 5963, 2551, 3368, 25, 13325, 11, 296, 25, 612, 14149, 3413, 8, 1492, 5468, 2122, 72798, 341, 262, 296, 1388, 4895, 28738, 5501, 33, 1777, 13867, 38545, 262, 7777, 51532, 633, 13657, 14401, 8793, 5680, 2658, 7177, 341, 262, 1005, 23941, 66, 9127, 487, 82956, 19004, 439, 10751, 401, 262, 1005, 18187, 487, 3867, 14784, 17489, 401, 262, 11765, 1985, 933, 262, 5279, 1633, 31115, 4552, 368, 341, 286, 1095, 5318, 21467, 284, 10751, 487, 2309, 545, 286, 21467, 7175, 1921, 1, 370, 3343, 998, 13554, 1535, 220, 16, 317, 286, 21467, 7175, 1921, 1, 4484, 3343, 998, 13554, 1535, 220, 17, 629, 286, 1095, 594, 284, 5027, 14784, 17489, 1921, 1, 69744, 498, 612, 81, 4129, 317, 286, 2105, 10945, 10509, 417, 11, 7638, 21362, 65, 1, 370, 498, 293, 1, 4484, 15399, 262, 457, 534]
```
Comment (jinu-jang) [inline tokenizer_ts/test/testdata/tokens.json:1]: Rename file to `tokens_gpt_3.5_turbo.json` for clarity

**Review by jinu-jang — COMMENTED** [review]
Typescript side reviewed.
Require more explanation on the core logic of the tokenization process.

**Review by jinu-jang — COMMENTED** [review]
Typescript side reviewed.
Require more explanation on the core logic of the tokenization process.

File: tokenizer_ts/src/tokenizerBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  let fileName = getFilenameFromUrl(mergeableRanksFileUrl);
+  await fetch(mergeableRanksFileUrl).then(response => {
+    if (response.ok) {
+      return response.text();
+    } else {
+      throw new Error(`Failed to fetch file from ${mergeableRanksFileUrl}. Status code: ${response.status}`);
+    }
+  })
+    .then(text => {
+      fs.writeFileSync(fileName, text);
```
Comment (jinu-jang) [inline tokenizer_ts/src/tokenizerBuilder.ts:191]: This just creates artifacts in the root directory of the project. At least route them to a `bin/` folder.

File: tokenizer_ts/src/tokenizerBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    default:
+      throw new Error(`Doesn't support this encoder [${encoderName}]`);
+  }
+
+  if (extraSpecialTokens !== null) {
+    specialTokens = new Map([...specialTokens, ...extraSpecialTokens]);
+  }
+
+  let fileName = getFilenameFromUrl(mergeableRanksFileUrl);
+  await fetch(mergeableRanksFileUrl).then(response => {
```
Comment (jinu-jang) [inline tokenizer_ts/src/tokenizerBuilder.ts:183]: No need to use Promise chaining when using async/await.
Consider following snippet.

```typescript
async function fetchAndSaveFile(mergeableRanksFileUrl: string, fileName: string): {
    const response = await fetch(mergeableRanksFileUrl);

    if (!response.ok) {
      throw new Error(`Failed to fetch file from ${mergeableRanksFileUrl}. Status code: ${response.status}`);
    }

    const text = response.text();
    fs.writeFileSync(fileName, text);
}
```

**Review by jinu-jang — COMMENTED** [review]
Forgot to talk about the tokenizer artifacts

File: tokenizer_ts/src/bytePairEncode.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  return str;
+}
+
+/** 
+ * This function implements the byte pair encoding algorithm.
+ * @param mergingBytes: bytes to be merged
+ * @param ranks: BPE rank for the bytes
+ * @returns number[]: Encoded token ids
+*/
+export function bytePairEncode(mergingBytes: Uint8Array, ranks: ReadonlyMap<string, number>): number[] {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/bytePairEncode.ts]: This is the same as in the rust implementation in tiktoken lib. Also a translation of the C# implementation. I don't want to take any unnecessary dependecy.

**Comment by shengyfu** [comment]
> Forgot to talk about the tokenizer artifacts

I will do npm publishing in a separate PR.

File: tokenizer_ts/test/tikTokenizer_p50k_base.test.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+const IM_START = "<|im_start|>";
+const IM_END = "<|im_end|>";
+const specialTokens: ReadonlyMap<string, number> = new Map([
+    [IM_START, 100264],
+    [IM_END, 100265],
+]);
+
+suite('TikTokenizer p50k_base Test Suite', function () {
+    let tokenizer_p50k_base: TikTokenizer;
+    beforeEach(async () => {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/test/tikTokenizer_p50k_base.test.ts]: Tried before, doesn't work.

File: tokenizer_ts/debug.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/debug.ts]: This is more for my own debugging purposes, I can leave it and the .vscode configurations out.

File: tokenizer_ts/src/tokenizerBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  }
+  return encoder;
+}
+
+/**
+ * Get the special tokens from the encoder name
+ * @param encoder encoder name
+ * @returns Map<string, number> special tokens mapping
+ */
+export function getSpecialTokensByEncoder(encoder: string): Map<string, number> {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/tokenizerBuilder.ts]: I exported MODEL_TO_ENODING map.

File: tokenizer_ts/src/tikTokenizer.ts [hunk 11]
```diff
@@ -0,0 +1,383 @@
+// Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/tikTokenizer.ts:1]: sure can, the purpose here is to create typescript version with no native dependency.

File: tokenizer_ts/src/tikTokenizer.ts [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+    }
+
+    /**
+     * Encode a piece of text limited by max token count through trimming suffix
+     * @param text text to encode
+     * @param maxTokenCount max token count
+     * @param allowedSpecial allowed special tokens
+     * @returns { tokenIds: number[], text: string } encoded token ids and trimmed text
+     */
+    public encodeTrimSuffix(text: string, maxTokenCount: number, allowedSpecial: ReadonlyArray<string>): { tokenIds: number[], text: string } {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/tikTokenizer.ts]: Most use cases will be calling encode, mixing the trimming cases with the basic encode function will make the code even less readable.

File: tokenizer_ts/src/bytePairEncode.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/**
+ * Convert a Uint8Array to a string
+ * @param uint8Array 
+ * @returns string
+ */
+export function uint8ArrayToString(uint8Array: Uint8Array): string {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/bytePairEncode.ts:9]: it's referenced in tikTokenizer.ts.

[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/bytePairEncode.ts:9]: yes, it is.

File: tokenizer_ts/src/tikTokenizer.ts [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+    /**
+     * Take the encoder tokens mapping from OpenAI tiktoken dump to build the encoder
+     * For gpt-3.5-turbo/gpt4, you can download the BPE tokens mapping from:
+     * https://openaipublic.blob.core.windows.net/encodings/cl100k_base.tiktoken
+     * @param tikTokenBpeFile BPE rank file path
+     * @param specialTokensEncoder special tokens encoder
+     * @param regexPattern regex pattern to split the input text
+     * @param cacheSize cache size
+     */
+    constructor(tikTokenBpeFile: string, specialTokensEncoder: ReadonlyMap<string, number>, regexPattern: string, cacheSize: number = 8192) {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/tikTokenizer.ts]: We can chat offline on what exactly you are looking for. Maybe address this in a later pr.

File: tokenizer_ts/src/tokenizerBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+  return createByEncoderName(getEncoderFromModelName(modelName), extraSpecialTokens);
+}
+
+/**
+ * Create a tokenizer from an encoder name
+ * @param encoderName encoder name
+ * @param extraSpecialTokens extra special tokens 
+ * @returns TikTokenizer tokenizer
+ */
+export async function createByEncoderName(encoderName: string, extraSpecialTokens: ReadonlyMap<string, number> | null = null): Promise<TikTokenizer> {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/tokenizerBuilder.ts]: This is matching OpenAI's implementation. I assume the user of the tokenizer knows how to set special tokens.

File: tokenizer_ts/src/tokenizerBuilder.ts [hunk 16]
```diff
@@ -0,0 +1,208 @@
+// Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/src/tokenizerBuilder.ts:1]: Exported MODEL_TO_ENCODING map.

File: tokenizer_ts/test/tikTokenizer.test.ts [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    assert.deepStrictEqual(encoded.tokenIds, [100264, 9906, 4435, 100265]);
+    assert.deepStrictEqual(encoded.text, str);
+
+    encoded = tokenizer.encodeTrimSuffix(str, 3, Array.from(specialTokens.keys()));
+    assert.deepStrictEqual(encoded.tokenIds, [100264, 9906, 4435]);
+    assert.deepStrictEqual(encoded.text, encodedStr);
+
+  });
+
+  test('encode trim suffix - 2', () => {
```
[the remark this answers is not in the document]
Comment (shengyfu) [reply tokenizer_ts/test/tikTokenizer.test.ts]: Fixed.

**Review by jinu-jang — APPROVED** [review]
Farther code improvements can be done in a separate PR.
Such as the failable constructor and other nitpicky changes.

**Pull request merged**